### PR TITLE
Short-circuit rebalancing when disabled

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today if `cluster.routing.rebalance.enable: none` then rebalancing is disabled,
but we still execute `balanceByWeights()` and perform some rather expensive
calculations before discovering that we cannot rebalance any shards. In a large
cluster this can make cluster state updates occur rather slowly. With this
change we check earlier whether rebalancing is globally disabled and, if so,
avoid the rebalancing process entirely.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+            .addAsNew(metaData.index("test"))
+            .build();
+
+        ClusterState clusterState = ClusterState.builder(ClusterName.CLUSTER_NAME_SETTING.getDefault(Settings.EMPTY))
+            .metaData(metaData).routingTable(routingTable).nodes(DiscoveryNodes.builder().add(newNode("node1"))).build();
+
+        allocationService.reroute(clusterState, "reroute").routingTable();
+    }
+
+    private AllocationService createAllocationService(Settings.Builder settings, ClusterPlugin plugin) {
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java]: NIT: can be `static`

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        final ClusterSettings emptyClusterSettings = new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS);
+        List<AllocationDecider> deciders = new ArrayList<>(ClusterModule.createAllocationDeciders(settings.build(), emptyClusterSettings,
+                Collections.singletonList(plugin)));
+        return new MockAllocationService(
+            new AllocationDeciders(deciders),
+            new TestGatewayAllocator(), new BalancedShardsAllocator(Settings.EMPTY), EmptyClusterInfoService.INSTANCE);
+    }
+
+    private class RebalanceShortCircuitPlugin implements ClusterPlugin {
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java]: NIT: can be `static`

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+            }
+
+            @Override
+            public Decision canRebalance(RoutingAllocation allocation) {
+                throw new AssertionError("canRebalance was not bypassed");
+            }
+        }
+    }
+
+    private class AllocateShortCircuitPlugin implements ClusterPlugin {
```
Comment (original-brownbear) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java]: NIT: can be `static`

**Review by original-brownbear — APPROVED** [review]
LGTM :)

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java]: Thanks, fixed in 149ff95

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java]: Thanks, fixed in 149ff95

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java]: Thanks, fixed in 149ff95

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+public class EnableAllocationShortCircuitTests extends ESAllocationTestCase {
+
+    public void testRebalancingSkippedIfDisabled() {
+        AllocationService allocationService = createAllocationService(Settings.builder()
+                .put(CLUSTER_ROUTING_REBALANCE_ENABLE_SETTING.getKey(), EnableAllocationDecider.Allocation.NONE.name()),
+            new RebalanceShortCircuitPlugin());
+
+        ClusterState clusterState = ClusterState.builder(ClusterName.CLUSTER_NAME_SETTING.get(Settings.EMPTY))
+            .nodes(DiscoveryNodes.builder().add(newNode("node1"))).build();
+
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java]: I think it makes sense to also add some indices in this test case?

**Review by henningandersen — CHANGES_REQUESTED** [review]
Looking good, I added a few smaller things to address.

File: server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationDecider.java [hunk 5]
```diff
@@ -136,10 +141,29 @@ public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, Routing
...
     }
 
+    @Override
+    public Decision canRebalance(RoutingAllocation allocation) {
+        if (allocation.ignoreDisable()) {
+            return allocation.decision(Decision.YES, NAME, "allocation is explicitly ignoring any disabling of rebalancing");
+        }
+
+        for (IndexMetaData indexMetaData : allocation.metaData()) {
+            if (INDEX_ROUTING_REBALANCE_ENABLE_SETTING.exists(indexMetaData.getSettings())) {
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationDecider.java]: Ok, I added this in 0c5c04b. I don't think this is a common configuration, so it'd be ok either way.

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java (hunk 4 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/EnableAllocationShortCircuitTests.java]: Sort of, although then we have to allocate their shards or else the `ClusterRebalanceAllocationDecider` steps in and prevents rebalancing. I expanded the tests to do this.

**Comment by DaveCTurner** [comment]
@elasticmachine run elasticsearch-ci/2

**Review by henningandersen — APPROVED** [review]
LGTM.

Thanks @DaveCTurner

**Pull request merged**